File: src/main/java/com/example/demo/controller/WebAuthnController.java
package com.example.demo.controller;

import com.example.demo.dto.WebAuthnAllowCredential;
import com.example.demo.dto.WebAuthnAuthenticationOptionsResponse;
import com.example.demo.dto.WebAuthnDiscoverableAuthenticationOptionsResponse;
import com.example.demo.dto.WebAuthnFinishAuthenticationRequest;
import com.example.demo.dto.WebAuthnFinishRegistrationRequest;
import com.example.demo.dto.WebAuthnPubKeyParam;
import com.example.demo.dto.WebAuthnRegistrationOptionsResponse;
import com.example.demo.dto.WebAuthnRp;
import com.example.demo.dto.WebAuthnUser;
import com.example.demo.model.WebAuthnCredential;
import com.example.demo.service.WebAuthnService;
import com.webauthn4j.util.Base64UrlUtil;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.web.authentication.session.SessionAuthenticationStrategy;
import org.springframework.security.web.context.HttpSessionSecurityContextRepository;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.nio.charset.StandardCharsets;
import java.util.*;

@Slf4j
@RestController
@RequestMapping("/api/webauthn")
@Validated
@RequiredArgsConstructor
public class WebAuthnController {

    private final WebAuthnService webAuthnService;
    private final UserDetailsService userDetailsService;
    private final SessionAuthenticationStrategy sessionAuthenticationStrategy;


    @Value("${webauthn.rp.id:localhost}")
    private String rpId;

    @Value("${webauthn.rp.name:Dev RP}")
    private String rpName;


    @GetMapping("/registration/options")
    public ResponseEntity<Object> registrationOptions(@RequestParam @NotBlank String username) {
        try {
            userDetailsService.loadUserByUsername(username);
        } catch (UsernameNotFoundException e) {
            log.warn("Registration attempted for non-existent user: {}", username);
            return ResponseEntity.badRequest().body(Map.of("message", "User not found"));
        }

        byte[] challenge = webAuthnService.generateChallenge();
        webAuthnService.saveRegistrationChallenge(username, challenge);

        WebAuthnRp rp = new WebAuthnRp(rpId, rpName);
        WebAuthnUser user = new WebAuthnUser(
                Base64UrlUtil.encodeToString(username.getBytes(StandardCharsets.UTF_8)),
                username,
                username
        );
        List<WebAuthnPubKeyParam> params = List.of(
                new WebAuthnPubKeyParam("public-key", -7),
                new WebAuthnPubKeyParam("public-key", -257)
        );
        WebAuthnRegistrationOptionsResponse res = new WebAuthnRegistrationOptionsResponse();
        res.setChallenge(Base64UrlUtil.encodeToString(challenge));
        res.setRp(rp);
        res.setUser(user);
        res.setPubKeyCredParams(params);
        return ResponseEntity.ok(res);
    }

    @PostMapping("/registration/finish")
    public ResponseEntity<Object> finishRegistration(
            @RequestParam @NotBlank String username,
            @Valid @RequestBody WebAuthnFinishRegistrationRequest req
    ) {
        byte[] challenge = webAuthnService.consumeRegistrationChallenge(username);
        if (challenge == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "Challenge not found or expired"));
        }

        try {
            byte[] clientDataJSON = Base64UrlUtil.decode(req.getResponse().getClientDataJSON());
            byte[] attestationObject = Base64UrlUtil.decode(req.getResponse().getAttestationObject());
            webAuthnService.registerCredential(username, challenge, clientDataJSON, attestationObject);
            log.info("Passkey registered successfully for user: {}", username);
            return ResponseEntity.ok(Map.of("message", "registered"));
        } catch (Exception e) {
            log.error("Failed to register passkey for user {}: {}", username, e.getMessage(), e);
            return ResponseEntity.badRequest().body(Map.of("message", "Registration failed: " + e.getMessage()));
        }
    }

    @GetMapping("/authentication/options")
    public ResponseEntity<Object> authenticationOptions(@RequestParam(required = false) String username) {
        byte[] challenge = webAuthnService.generateChallenge();
        if (username == null || username.isBlank()) {
            String challengeId = webAuthnService.saveDiscoverableChallenge(challenge);
            WebAuthnDiscoverableAuthenticationOptionsResponse res = new WebAuthnDiscoverableAuthenticationOptionsResponse();
            res.setChallenge(Base64UrlUtil.encodeToString(challenge));
            res.setRpId(rpId);
            Map<String, Object> body = new HashMap<>();
            body.put("challenge", res.getChallenge());
            body.put("rpId", res.getRpId());
            body.put("timeout", res.getTimeout());
            body.put("userVerification", res.getUserVerification());
            body.put("challengeId", challengeId);
            return ResponseEntity.ok(body);
        }

        List<WebAuthnCredential> creds = webAuthnService.findCredentials(username);
        if (creds.isEmpty()) {
            return ResponseEntity.status(404).body(Map.of("message", "No passkeys registered for this user"));
        }

        webAuthnService.saveAuthenticationChallenge(username, challenge);

        List<WebAuthnAllowCredential> allow = new ArrayList<>();
        for (WebAuthnCredential cred : creds) {
            List<String> transports = new ArrayList<>();
            if (cred.getTransports() != null) {
                Collections.addAll(transports, cred.getTransports().split(","));
            }
            allow.add(new WebAuthnAllowCredential("public-key", Base64UrlUtil.encodeToString(cred.getCredentialId()), transports));
        }
        WebAuthnAuthenticationOptionsResponse res = new WebAuthnAuthenticationOptionsResponse();
        res.setChallenge(Base64UrlUtil.encodeToString(challenge));
        res.setRpId(rpId);
        res.setAllowCredentials(allow);
        return ResponseEntity.ok(res);
    }

    @PostMapping("/authentication/finish/discoverable")
    public ResponseEntity<Object> finishDiscoverableAuthentication(
            @RequestParam @NotBlank String challengeId,
            @Valid @RequestBody WebAuthnFinishAuthenticationRequest req,
            HttpServletRequest request,
            HttpServletResponse response
    ) {
        byte[] challenge = webAuthnService.consumeDiscoverableChallenge(challengeId);
        if (challenge == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "Challenge not found or expired"));
        }

        byte[] credentialId = Base64UrlUtil.decode(req.getRawId());
        WebAuthnCredential credential = webAuthnService.findByCredentialId(credentialId);
        if (credential == null) {
            return ResponseEntity.status(404).body(Map.of("message", "Credential not found"));
        }

        String userHandle = req.getResponse().getUserHandle();
        if (userHandle == null || userHandle.isBlank()) {
            return ResponseEntity.badRequest().body(Map.of("message", "userHandle is required for discoverable credentials"));
        }

        String usernameFromHandle = new String(Base64UrlUtil.decode(userHandle), StandardCharsets.UTF_8);
        if (!credential.getUsername().equals(usernameFromHandle)) {
            log.warn("Credential ownership mismatch: credential belongs to {}, but userHandle indicates {}",
                    credential.getUsername(), usernameFromHandle);
            return ResponseEntity.status(403).body(Map.of("message", "Credential does not match userHandle"));
        }

        try {
            byte[] clientDataJSON = Base64UrlUtil.decode(req.getResponse().getClientDataJSON());
            byte[] authenticatorData = Base64UrlUtil.decode(req.getResponse().getAuthenticatorData());
            byte[] signature = Base64UrlUtil.decode(req.getResponse().getSignature());
            byte[] userHandleBytes = Base64UrlUtil.decode(userHandle);

            webAuthnService.verifyAssertion(
                    credential,
                    challenge,
                    credentialId,
                    clientDataJSON,
                    authenticatorData,
                    signature,
                    userHandleBytes
            );

            establishSecuritySession(credential.getUsername(), request, response);

            log.info("Discoverable passkey authentication successful for user: {}", credential.getUsername());
            return ResponseEntity.ok(Map.of("message", "authenticated", "username", credential.getUsername()));
        } catch (Exception e) {
            log.error("Discoverable passkey authentication failed: {} - {}", e.getClass().getSimpleName(), e.getMessage(), e);
            String errorMessage = e.getMessage() != null ? e.getMessage() : e.getClass().getSimpleName();
            return ResponseEntity.status(401).body(Map.of("message", "Authentication failed: " + errorMessage));
        }
    }

    @PostMapping("/authentication/finish")
    public ResponseEntity<Object> finishAuthentication(
            @RequestParam @NotBlank String username,
            @Valid @RequestBody WebAuthnFinishAuthenticationRequest req,
            HttpServletRequest request,
            HttpServletResponse response
    ) {
        byte[] challenge = webAuthnService.consumeAuthenticationChallenge(username);
        if (challenge == null) {
            return ResponseEntity.badRequest().body(Map.of("message", "Challenge not found or expired"));
        }

        byte[] credentialId = Base64UrlUtil.decode(req.getRawId());
        WebAuthnCredential credential = webAuthnService.findByCredentialId(credentialId);
        if (credential == null) {
            return ResponseEntity.status(404).body(Map.of("message", "Credential not found"));
        }

        if (!credential.getUsername().equals(username)) {
            log.warn("Credential ownership mismatch: expected {}, got {}", username, credential.getUsername());
            return ResponseEntity.status(403).body(Map.of("message", "Credential does not belong to user"));
        }

        try {
            byte[] clientDataJSON = Base64UrlUtil.decode(req.getResponse().getClientDataJSON());
            byte[] authenticatorData = Base64UrlUtil.decode(req.getResponse().getAuthenticatorData());
            byte[] signature = Base64UrlUtil.decode(req.getResponse().getSignature());
            byte[] userHandle = req.getResponse().getUserHandle() != null
                    ? Base64UrlUtil.decode(req.getResponse().getUserHandle())
                    : null;

            webAuthnService.verifyAssertion(
                    credential,
                    challenge,
                    credentialId,
                    clientDataJSON,
                    authenticatorData,
                    signature,
                    userHandle
            );

            establishSecuritySession(username, request, response);

            log.info("Passkey authentication successful for user: {}", username);
            return ResponseEntity.ok(Map.of("message", "authenticated"));
        } catch (Exception e) {
            log.error("Passkey authentication failed for user {}: {} - {}", username, e.getClass().getSimpleName(), e.getMessage(), e);
            String errorMessage = e.getMessage() != null ? e.getMessage() : e.getClass().getSimpleName();
            return ResponseEntity.status(401).body(Map.of("message", "Authentication failed: " + errorMessage));
        }
    }

    private void establishSecuritySession(String username, HttpServletRequest request, HttpServletResponse response) {
        var userDetails = userDetailsService.loadUserByUsername(username);
        var authentication = new UsernamePasswordAuthenticationToken(userDetails, null, userDetails.getAuthorities());
        var securityContext = SecurityContextHolder.createEmptyContext();
        securityContext.setAuthentication(authentication);
        SecurityContextHolder.setContext(securityContext);

        sessionAuthenticationStrategy.onAuthentication(authentication, request, response);

        var session = request.getSession(true);
        session.setAttribute(HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY, securityContext);
    }

    @GetMapping("/credentials")
    public ResponseEntity<Object> listCredentials(@RequestParam @NotBlank String username) {
        List<WebAuthnCredential> creds = webAuthnService.findCredentials(username);
        List<Map<String, Object>> list = new ArrayList<>();
        for (WebAuthnCredential cred : creds) {
            Map<String, Object> item = new HashMap<>();
            item.put("id", cred.getId());
            item.put("credentialId", Base64UrlUtil.encodeToString(cred.getCredentialId()));
            item.put("createdAt", cred.getCreatedAt() != null ? cred.getCreatedAt().toString() : null);
            item.put("isValid", cred.getPublicKeyCose() != null && cred.getPublicKeyCose().length > 0);
            list.add(item);
        }
        return ResponseEntity.ok(Map.of("credentials", list));
    }

    @DeleteMapping("/credentials/{id}")
    public ResponseEntity<Object> deleteCredential(@PathVariable Long id, @RequestParam @NotBlank String username) {
        WebAuthnCredential credential = webAuthnService.findCredentialById(id);
        if (credential == null) {
            return ResponseEntity.status(404).body(Map.of("message", "Credential not found"));
        }

        if (!credential.getUsername().equals(username)) {
            return ResponseEntity.status(403).body(Map.of("message", "Credential does not belong to user"));
        }

        webAuthnService.deleteCredential(id);
        log.info("Passkey deleted for user: {}, credentialId: {}", username, id);
        return ResponseEntity.ok(Map.of("message", "Credential deleted"));
    }

}

File: src/main/kotlin/com/example/demo/controller/WebAuthnController.kt
package com.example.demo.controller

import com.example.demo.service.WebAuthnService
import com.webauthn4j.util.Base64UrlUtil
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.web.authentication.session.SessionAuthenticationStrategy
import org.springframework.security.web.context.HttpSessionSecurityContextRepository
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/webauthn")
class WebAuthnController(
    private val webAuthnService: WebAuthnService,
    private val userDetailsService: UserDetailsService,
    private val sessionAuthenticationStrategy: SessionAuthenticationStrategy
) {
    private val logger = LoggerFactory.getLogger(WebAuthnController::class.java)

    @Value("\${webauthn.rp.id:localhost}")
    private lateinit var rpId: String

    @Value("\${webauthn.rp.name:Dev RP}")
    private lateinit var rpName: String

    data class RegistrationOptionsResponse(
        val challenge: String,
        val rp: Rp,
        val user: User,
        val pubKeyCredParams: List<PubKeyParam>,
        val timeout: Long = 60000,
        val attestation: String = "none",
        val authenticatorSelection: AuthenticatorSelection = AuthenticatorSelection()
    )

    data class Rp(val id: String, val name: String)
    data class User(val id: String, val name: String, val displayName: String)
    data class PubKeyParam(val type: String, val alg: Int)
    data class AuthenticatorSelection(
        val authenticatorAttachment: String? = null,
        val residentKey: String = "required",  // Discoverable Credential必須
        val requireResidentKey: Boolean = true,
        val userVerification: String = "preferred"
    )

    data class FinishRegistrationRequest(
        val id: String,
        val rawId: String,
        val type: String,
        val response: RegistrationResponse
    )

    data class RegistrationResponse(
        val attestationObject: String,
        val clientDataJSON: String
    )

    // ユーザー名なし認証用のレスポンス（allowCredentials空）
    data class DiscoverableAuthenticationOptionsResponse(
        val challenge: String,
        val rpId: String,
        val timeout: Long = 60000,
        val userVerification: String = "preferred"
    )

    data class AuthenticationOptionsResponse(
        val challenge: String,
        val rpId: String,
        val allowCredentials: List<AllowCredential>,
        val timeout: Long = 60000,
        val userVerification: String = "preferred"
    )

    data class AllowCredential(val type: String, val id: String, val transports: List<String> = emptyList())

    data class FinishAuthenticationRequest(
        val id: String,
        val rawId: String,
        val type: String,
        val response: AuthenticationResponse
    )

    data class AuthenticationResponse(
        val authenticatorData: String,
        val clientDataJSON: String,
        val signature: String,
        val userHandle: String?
    )

    @GetMapping("/registration/options")
    fun registrationOptions(@RequestParam username: String): ResponseEntity<Any> {
        // ユーザーが存在するか確認
        try {
            userDetailsService.loadUserByUsername(username)
        } catch (e: UsernameNotFoundException) {
            logger.warn("Registration attempted for non-existent user: $username")
            return ResponseEntity.badRequest().body(mapOf("message" to "User not found"))
        }

        val challenge = webAuthnService.generateChallenge()
        webAuthnService.saveRegistrationChallenge(username, challenge)

        val rp = Rp(id = rpId, name = rpName)
        val user = User(
            id = Base64UrlUtil.encodeToString(username.toByteArray()),
            name = username,
            displayName = username
        )
        val params = listOf(
            PubKeyParam(type = "public-key", alg = -7),   // ES256
            PubKeyParam(type = "public-key", alg = -257)  // RS256
        )
        val res = RegistrationOptionsResponse(
            challenge = Base64UrlUtil.encodeToString(challenge),
            rp = rp,
            user = user,
            pubKeyCredParams = params
        )
        return ResponseEntity.ok(res)
    }

    @PostMapping("/registration/finish")
    fun finishRegistration(
        @RequestParam username: String,
        @RequestBody req: FinishRegistrationRequest
    ): ResponseEntity<Any> {
        val challenge = webAuthnService.consumeRegistrationChallenge(username)
            ?: return ResponseEntity.badRequest().body(mapOf("message" to "Challenge not found or expired"))

        return try {
            val clientDataJSON = Base64UrlUtil.decode(req.response.clientDataJSON)
            val attestationObject = Base64UrlUtil.decode(req.response.attestationObject)

            webAuthnService.registerCredential(
                username = username,
                challenge = challenge,
                clientDataJSON = clientDataJSON,
                attestationObject = attestationObject
            )
            logger.info("Passkey registered successfully for user: $username")
            ResponseEntity.ok(mapOf("message" to "registered"))
        } catch (e: Exception) {
            logger.error("Failed to register passkey for user $username: ${e.message}", e)
            ResponseEntity.badRequest().body(mapOf("message" to "Registration failed: ${e.message}"))
        }
    }

    @GetMapping("/authentication/options")
    fun authenticationOptions(@RequestParam(required = false) username: String?): ResponseEntity<Any> {
        val challenge = webAuthnService.generateChallenge()

        // ユーザー名が指定されていない場合はDiscoverable Credential認証
        if (username.isNullOrBlank()) {
            // ユーザー名なしの場合、チャレンジをセッションIDベースで保存
            val challengeId = webAuthnService.saveDiscoverableChallenge(challenge)
            val res = DiscoverableAuthenticationOptionsResponse(
                challenge = Base64UrlUtil.encodeToString(challenge),
                rpId = rpId
            )
            return ResponseEntity.ok(mapOf(
                "challenge" to res.challenge,
                "rpId" to res.rpId,
                "timeout" to res.timeout,
                "userVerification" to res.userVerification,
                "challengeId" to challengeId
            ))
        }

        // ユーザー名が指定されている場合は従来の認証
        val creds = webAuthnService.findCredentials(username)
        if (creds.isEmpty()) {
            return ResponseEntity.status(404).body(mapOf("message" to "No passkeys registered for this user"))
        }

        webAuthnService.saveAuthenticationChallenge(username, challenge)

        val allow = creds.map {
            AllowCredential(
                type = "public-key",
                id = Base64UrlUtil.encodeToString(it.credentialId),
                transports = it.transports?.split(",") ?: emptyList()
            )
        }
        val res = AuthenticationOptionsResponse(
            challenge = Base64UrlUtil.encodeToString(challenge),
            rpId = rpId,
            allowCredentials = allow
        )
        return ResponseEntity.ok(res)
    }

    /**
     * ユーザー名なしパスキー認証（Discoverable Credentials）
     */
    @PostMapping("/authentication/finish/discoverable")
    fun finishDiscoverableAuthentication(
        @RequestParam challengeId: String,
        @RequestBody req: FinishAuthenticationRequest,
        request: HttpServletRequest,
        response: HttpServletResponse
    ): ResponseEntity<Any> {
        val challenge = webAuthnService.consumeDiscoverableChallenge(challengeId)
            ?: return ResponseEntity.badRequest().body(mapOf("message" to "Challenge not found or expired"))

        val credentialId = Base64UrlUtil.decode(req.rawId)
        val credential = webAuthnService.findByCredentialId(credentialId)
            ?: return ResponseEntity.status(404).body(mapOf("message" to "Credential not found"))

        // userHandleからユーザー名を取得して検証
        val userHandle = req.response.userHandle
        if (userHandle.isNullOrBlank()) {
            return ResponseEntity.badRequest().body(mapOf("message" to "userHandle is required for discoverable credentials"))
        }

        val usernameFromHandle = String(Base64UrlUtil.decode(userHandle), Charsets.UTF_8)
        if (credential.username != usernameFromHandle) {
            logger.warn("Credential ownership mismatch: credential belongs to ${credential.username}, but userHandle indicates $usernameFromHandle")
            return ResponseEntity.status(403).body(mapOf("message" to "Credential does not match userHandle"))
        }

        return try {
            val clientDataJSON = Base64UrlUtil.decode(req.response.clientDataJSON)
            val authenticatorData = Base64UrlUtil.decode(req.response.authenticatorData)
            val signature = Base64UrlUtil.decode(req.response.signature)
            val userHandleBytes = Base64UrlUtil.decode(userHandle)

            // WebAuthn署名検証
            webAuthnService.verifyAssertion(
                credential = credential,
                challenge = challenge,
                credentialId = credentialId,
                clientDataJSON = clientDataJSON,
                authenticatorData = authenticatorData,
                signature = signature,
                userHandle = userHandleBytes
            )

            // Spring Securityの認証セッションを確立
            establishSecuritySession(credential.username, request, response)

            logger.info("Discoverable passkey authentication successful for user: ${credential.username}")
            ResponseEntity.ok(mapOf("message" to "authenticated", "username" to credential.username))
        } catch (e: Exception) {
            logger.error("Discoverable passkey authentication failed: ${e::class.simpleName} - ${e.message}", e)
            val errorMessage = e.message ?: e::class.simpleName ?: "Unknown error"
            ResponseEntity.status(401).body(mapOf("message" to "Authentication failed: $errorMessage"))
        }
    }

    @PostMapping("/authentication/finish")
    fun finishAuthentication(
        @RequestParam username: String,
        @RequestBody req: FinishAuthenticationRequest,
        request: HttpServletRequest,
        response: HttpServletResponse
    ): ResponseEntity<Any> {
        val challenge = webAuthnService.consumeAuthenticationChallenge(username)
            ?: return ResponseEntity.badRequest().body(mapOf("message" to "Challenge not found or expired"))

        val credentialId = Base64UrlUtil.decode(req.rawId)
        val credential = webAuthnService.findByCredentialId(credentialId)
            ?: return ResponseEntity.status(404).body(mapOf("message" to "Credential not found"))

        // クレデンシャルが指定されたユーザーに属しているか確認
        if (credential.username != username) {
            logger.warn("Credential ownership mismatch: expected $username, got ${credential.username}")
            return ResponseEntity.status(403).body(mapOf("message" to "Credential does not belong to user"))
        }

        return try {
            val clientDataJSON = Base64UrlUtil.decode(req.response.clientDataJSON)
            val authenticatorData = Base64UrlUtil.decode(req.response.authenticatorData)
            val signature = Base64UrlUtil.decode(req.response.signature)
            val userHandle = req.response.userHandle?.let { Base64UrlUtil.decode(it) }

            // WebAuthn署名検証
            webAuthnService.verifyAssertion(
                credential = credential,
                challenge = challenge,
                credentialId = credentialId,
                clientDataJSON = clientDataJSON,
                authenticatorData = authenticatorData,
                signature = signature,
                userHandle = userHandle
            )

            // Spring Securityの認証セッションを確立
            establishSecuritySession(username, request, response)

            logger.info("Passkey authentication successful for user: $username")
            ResponseEntity.ok(mapOf("message" to "authenticated"))
        } catch (e: Exception) {
            logger.error("Passkey authentication failed for user $username: ${e::class.simpleName} - ${e.message}", e)
            val errorMessage = e.message ?: e::class.simpleName ?: "Unknown error"
            ResponseEntity.status(401).body(mapOf("message" to "Authentication failed: $errorMessage"))
        }
    }

    private fun establishSecuritySession(
        username: String,
        request: HttpServletRequest,
        response: HttpServletResponse
    ) {
        val userDetails = userDetailsService.loadUserByUsername(username)
        val authentication = UsernamePasswordAuthenticationToken(
            userDetails,
            null,
            userDetails.authorities
        )
        val securityContext = SecurityContextHolder.createEmptyContext()
        securityContext.authentication = authentication
        SecurityContextHolder.setContext(securityContext)

        // Spring Securityの標準セッション戦略を適用
        sessionAuthenticationStrategy.onAuthentication(authentication, request, response)

        // セッションにセキュリティコンテキストを保存
        val session = request.getSession(true)
        session.setAttribute(
            HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY,
            securityContext
        )
    }

    /**
     * ユーザーの登録済みパスキー一覧を取得
     */
    @GetMapping("/credentials")
    fun listCredentials(@RequestParam username: String): ResponseEntity<Any> {
        val creds = webAuthnService.findCredentials(username)
        val list = creds.map { cred ->
            mapOf(
                "id" to cred.id,
                "credentialId" to Base64UrlUtil.encodeToString(cred.credentialId),
                "createdAt" to cred.createdAt?.toString(),
                "isValid" to cred.publicKeyCose.isNotEmpty()
            )
        }
        return ResponseEntity.ok(mapOf("credentials" to list))
    }

    /**
     * 指定したパスキーを削除
     */
    @DeleteMapping("/credentials/{id}")
    fun deleteCredential(
        @PathVariable id: Long,
        @RequestParam username: String
    ): ResponseEntity<Any> {
        val credential = webAuthnService.findCredentialById(id)
            ?: return ResponseEntity.status(404).body(mapOf("message" to "Credential not found"))

        if (credential.username != username) {
            return ResponseEntity.status(403).body(mapOf("message" to "Credential does not belong to user"))
        }

        webAuthnService.deleteCredential(id)
        logger.info("Passkey deleted for user: $username, credentialId: $id")
        return ResponseEntity.ok(mapOf("message" to "Credential deleted"))
    }
}
